Undo/redo functionality addition

stutijain2006 opened the pull request:
Solves a long standing query of addition of Undo/Redo Feature 
Solves #4561 

This PR introduces undo/redo mechanism across various key interactions in music block ensuring common block operations can be safely reversed and reapplied. 

Implemented Features- 
1. Undo/Redo functionality for Block Deletion- When a block(or block stack) is deleted, the action can be undone to bring it back or redone to send it again to trash.
2. Undo/Redo functionality for block movement - The most important utilization of undo/redo being the movement of blocks, allowing the blocks (including entire connected stack) to return to its original positions . It restores the full stack layout not just the start position. 
3. Undo/Redo functionality for Addition of New Block - Undo /Redo can be applied when a new block is added , such that newly added blocks can be undone to remove them and redone to restore them with their original positions and  connections intact.


Feature can be accessed through two ways- 

1. Keyboard Shortcuts- Undo can be accessed through Ctrl+Z and Redo through Ctrl+Y 
2. Through tooltip buttons- There are designated buttons for Undo, Redo in the toolbar from where these functions can be accessed. 

Current Limitations and Future Enhancements -

1. Currently undo/redo is restricted only to new block addition, block deletion and block movement but this can further be enhanced to renaming the blocks, value changes (e.g number blocks) and to advanced block addition. 

- [x] Feature

stutijain2006 committed:
Add undo/redo feature in basic features like block creation, delete, movement

stutijain2006 committed:
Added undo/redo feature in complete block movement as stack, new blocks loaded and functionality of undo/redo buttons

zealot-zew reviewed (commented):
The bugs i encountered:
1) the trash can seems disabled (the trash doesn't appear in the trash can)
2) if i undo 'n' number of time (where n > the blocks i deleted) the workspace acts weirdly by removing the start and making a silence block appear. 
3) i like it how the undo feature brings back the block to its original place but it doesn't fit in the block (it just stays in the position in the background)

https://github.com/user-attachments/assets/d7f07da5-744a-4064-83ba-697c4831d5d7

stutijain2006 committed:
Updated undo/redo function to send the item to trash , connections on movement of a block

stutijain2006 commented:
> 3\. e block

I have implemented the required changes -
1. When a block is moved and Undo is performed, the block returns to its original position with all connections retained. 
2. When a block is deleted, it is sent to trash, from where it can be restored manually from restore from trash option or Undo can be performed.
3. Repeated Undo operations will no longer introduce extra elements like Silence block

stutijain2006 committed:
Changed the code for paret resizing while performing undo/redo for inner block movement

stutijain2006 commented:
Heyy @walterbender 
I am working on this parent sizing issue from a long time now, but unable to figure out how to resolve this issue, I tried -
1. Treating Start block as clamp/expandable block and tried routing it through adjustExpandableClampBlock
2. Implemented a function resizeStartBlock for manual resizing of Start block  but that caused some internal conflicts. 
3. Tried VspaceHandling for Start block by both allowing and removing vspace under Start block and even skipping addremoveVspaceBlock for Start, but the output was getStackSize was returning the correct values even when visuals don't update. 
4. Updated CachedHeight for Start block.
5. Rebuilt RestoreState function to match blockMoved function and rebuild clamp bottom-up

But the conclusion is none of this works. So I request you, if you can guide me how I can proceed and in which direction to resolve this issue would be grateful.

stutijain2006 committed:
Merge main branch of master into myfork/Undo/Redo

stutijain2006 committed:
Fixed parent sizing issue for block removal and undo

stutijain2006 committed:
Merge branch 'master' of https://github.com/sugarlabs/musicblocks into Undo/Redo

stutijain2006 committed:
Merge branch 'master' of https://github.com/sugarlabs/musicblocks into Undo/Redo

stutijain2006 committed:
Solved parent block resizing

stutijain2006 committed:
Applied prettier and removed console log

stutijain2006 committed:
Updated undo redo whitescreen issue and added test for the same

stutijain2006 committed:
Updated new block insertion to cover all the blocks available

stutijain2006 committed:
Merge branch 'master' of https://github.com/sugarlabs/musicblocks into Undo/Redo

stutijain2006 committed:
Fixed failing palette test, cypress test and lint errors

stutijain2006 committed:
Run prettier on failing files

vanshika2720 commented:
@stutijain2006 merge conflicts ....please take a look.

stutijain2006 committed:
Merge branch 'master' into Undo/Redo

stutijain2006 committed:
Applied prettier on blocks.js to solve es lint errors

stutijain2006 committed:
Merge branch 'master' into Undo/Redo
